sql/mysqld_thd_manager.cc

```diff
@@ -205,6 +206,18 @@ void Global_THD_manager::add_thd(THD *thd) {
 
 void Global_THD_manager::remove_thd(THD *thd) {
   DBUG_PRINT("info", ("Global_THD_manager::remove_thd %p", thd));
+
+  /*
+    It is possible when a slave connection dies after register_slave,
+    it may end up leaking SLAVE_INFO structure with a dangling THD pointer
+    In most cases it may be freed by the same slave with the same server_id
+    when it register itself again with the same server_id, but if the
+    server_id ends up being different the entry could be leaked forever and
+    with an invalid THD pointer
+    To address this issue, we let the THD remove itself from the slave list
+  */
+  unregister_slave(thd, true, true);
```

Comment:
This line seems to cause the merge_large_tests-t unit test to fail with:

```
[ RUN      ] ThreadManagerTest.AddRemoveTHDWithGuard 
merge_large_tests-t: /home/herman/rocks-mysql/8.0/include/thr_mutex.h:186: int my_mutex_lock(my_mutex_t *, const char *, uint): Assertion `mp->m_u.m_safe_ptr != __null' failed.
                                                    
Thread 1 "merge_large_tes" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt                                                                                                
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7a21935 in __GI_abort () at abort.c:90
#2  0x00007ffff7a30107 in __assert_fail_base (fmt=0x7ffff7ba2200 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x342c865 "mp->m_u.m_safe_ptr != __null", 
    file=file@entry=0x3a27fcf "/home/herman/rocks-mysql/8.0/include/thr_mutex.h", line=line@entry=186, function=function@entry=0x354ff6a "int my_mutex_lock(my_mutex_t *, const char *, uint)") at assert.c:92
#3  0x00007ffff7a30173 in __GI___assert_fail (assertion=0x342c865 "mp->m_u.m_safe_ptr != __null", file=0x3a27fcf "/home/herman/rocks-mysql/8.0/include/thr_mutex.h", line=186, 
    function=0x354ff6a "int my_mutex_lock(my_mutex_t *, const char *, uint)") at assert.c:101
#4  0x0000000005c76704 in my_mutex_lock (mp=0x8867f60 <LOCK_slave_list>, file=0x35581e7 "/home/herman/rocks-mysql/8.0/sql/rpl_master.cc", line=198) at /home/herman/rocks-mysql/8.0/include/thr_mutex.h:186
#5  0x0000000005c71676 in inline_mysql_mutex_lock (that=0x8867f60 <LOCK_slave_list>, src_file=0x35581e7 "/home/herman/rocks-mysql/8.0/sql/rpl_master.cc", src_line=198)
    at /home/herman/rocks-mysql/8.0/include/mysql/psi/mysql_mutex.h:267
#6  0x0000000005c716df in unregister_slave (thd=0x7fffffffa030, only_mine=true, need_lock_slave_list=true) at /home/herman/rocks-mysql/8.0/sql/rpl_master.cc:198
#7  0x0000000005637c70 in Global_THD_manager::remove_thd (this=0x7ffff52be000, thd=0x7fffffffa030) at /home/herman/rocks-mysql/8.0/sql/mysqld_thd_manager.cc:219
#8  0x000000000519138f in thd_manager_unittest::ThreadManagerTest_AddRemoveTHDWithGuard_Test::TestBody (this=0x7ffff53fc340) at /home/herman/rocks-mysql/8.0/unittest/gunit/thd_manager-t.cc:79
#9  0x000000000805c01e in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x7ffff53fc340, method=&virtual testing::Test::TestBody(), location=0x39f1894 "the test body")
    at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2417
#10 0x000000000804d6fe in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x7ffff53fc340, method=&virtual testing::Test::TestBody(), location=0x39f1894 "the test body")
    at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2453
#11 0x0000000008033b63 in testing::Test::Run (this=0x7ffff53fc340) at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2490
#12 0x00000000080349a3 in testing::TestInfo::Run (this=0x7ffff52976c0) at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2677
#13 0x000000000803507f in testing::TestCase::Run (this=0x7ffff529be00) at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2802
#14 0x000000000803d457 in testing::internal::UnitTestImpl::RunAllTests (this=0x7ffff522e000)
    at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:4737
#15 0x000000000805eb7e in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7ffff522e000, 
    method=(bool (testing::internal::UnitTestImpl::*)(class testing::internal::UnitTestImpl * const)) 0x803cfa0 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x30e88dd "auxiliary test code (environments or event listeners)") at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2417
#16 0x000000000804f744 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7ffff522e000, 
    method=(bool (testing::internal::UnitTestImpl::*)(class testing::internal::UnitTestImpl * const)) 0x803cfa0 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x30e88dd "auxiliary test code (environments or event listeners)") at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:2453
#17 0x000000000803cf53 in testing::UnitTest::Run (this=0x894b590 <testing::UnitTest::GetInstance()::instance>)
    at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/src/gtest.cc:4331
#18 0x0000000006ca7771 in RUN_ALL_TESTS () at /mnt/gvfs/third-party2/googletest/4e0d1647e5289c4a821397e0d728b8478b9546ce/master/src/googletest/googletest/include/gtest/gtest.h:2255
#19 0x0000000006ca7428 in main (argc=0, argv=0x7fffffffdbe0) at /home/herman/rocks-mysql/8.0/unittest/gunit/gunit_test_main_server.cc:91
```

Reply:
It's mainly because init_slave_list() is not called to initialize LOCK_slave_list before the test calls remove_thd() and unregister_slave().

Reply:
Fixed